src/host/alloc_cache.jl

```diff
@@ -0,0 +1,152 @@
+    Base.@lock cache.lock push!(busy_pool, x)
+    return x
+end
+
+function free_busy!(cache::AllocCache)
+    for uid in cache.busy.keys
+        busy_pool = get_pool!(cache, :busy, uid)
+        isempty(busy_pool) && continue
+
+        Base.@lock cache.lock begin
+            free_pool = get_pool!(cache, :free, uid)
+            append!(free_pool, busy_pool)
+            empty!(busy_pool)
+        end
+    end
+    return
+end
+
+function unsafe_free!(cache::AllocCache)
+    Base.@lock cache.lock begin
+        for (_, pool) in cache.busy
+            isempty(pool) || error(
+                "Invalidating allocations cache that's currently in use. " *
+                    "Invalidating inside `@enable` is not allowed."
+            )
+        end
+        for (_, pool) in cache.free
+            map(unsafe_free!, pool)
+        end
+        empty!(cache.free)
+    end
+    return
+end
+
+function Base.sizeof(cache::AllocCache)
+    sz = UInt64(0)
+    Base.@lock cache.lock begin
+        for kind in (cache.free, cache.busy), (_, pool) in kind
+            sz += sum(sizeof, pool; init = UInt64(0))
+        end
+    end
+    return sz
+end
+
+const ALLOC_CACHE = ScopedValue{Union{Nothing, AllocCache}}(nothing)
+
+"""
+    @enable(cache, expr)
+
+Evaluate expression `expr` using allocations cache `cache`.
+
+When gpu allocation is requested during execution of `expr`,
+it will first check if there's "free" cache instead of performing an actual allocation.
+If no "free" allocation exists, an actual allocation is performed.
+Before returning allocation to the user, it is marked as busy and
+will not be used by allocation in the scope defined by `@enable`.
+
+**After** the execution of `expr` all "busy" allocations are marked as "free"
+thus they can be re-used next time the program enters this scope.
+
+This is useful to apply in a repeating block of code to avoid relying on
+GC to free gpu memory in time.
+
+# Example
+
+In the following example, each iteration of the for-loop requires `8 GiB` of gpu memory.
+Without caching allocator GC wouldn't be able to free arrays in time
+resulting in higher memory usage.
+With caching allocator, memory usage stays at exactly `8 GiB`.
+
+```julia
+cache = GPUArrays.AllocCache(CuArray)
+n = 1024^3
+for i in 1:1000
+    GPUArrays.@enable cache begin
+        sin.(CUDA.rand(Float32, n))
+    end
+end
+# To free immediately.
+# Otherwise, it will be freed when collected by GC.
+GPUArrays.unsafe_free!(cache)
+```
+
+See [`@disable`](@ref).
+"""
+macro enable(cache, expr)
+    return quote
+        res = @with $(esc(ALLOC_CACHE)) => $(esc(cache)) $(esc(expr))
+        free_busy!($(esc(cache)))
+        res
+    end
+end
+
+"""
+    disable(expr)
+
+Evaluate expression `expr` without using allocations cache.
+This is useful to call from within `@enable` to avoid caching some allocations.
+"""
+macro disable(expr)
+    return quote
+        @with $(esc(ALLOC_CACHE)) => nothing $(esc(expr))
+    end
+end
```

Comment:
Seeing how these are in the main GPUArrays module, I'd opt for more descriptive names. `@cached` and `@uncached` maybe?

Alternatively, with do-block syntaxes always taking the `cache` argument the intent would be clear (`enable(cache) do ... end`, `disable(cache) do ... end`), but I take it you're purposefully avoiding closures so that you don't run into boxing issues?

Reply:
Renamed.

> but I take it you're purposefully avoiding closures so that you don't run into boxing issues?

With macro I like that you can define variables and don't need to return them as would be with the do-block syntax at the end.